### ISSUE TITLE
regridding method for comparison with observations

### ISSUE THREAD

**Issue by JavierDiezSierra** [euro-cordex/joint-evaluation#24]
We are going to need to regrid CORDEX data (only the climatologies, not the raw data) to compare the results with  observations/reanalysis. We implemented some [Python functions](https://github.com/SantanderMetGroup/python-resources/blob/main/lib/interpolater.py) for "conservative" remapping using the [xESMF](https://xesmf.readthedocs.io/en/stable/) package within the framework of the Copernicus Atlas.

 @larsbuntemeyer I mention this because I saw that you use the same Python package, and might be interesting.

@jesusff

**Comment by larsbuntemeyer** [comment]
Yes, with xESMF regridding becomes basically a one liner (if data complies with CF conventions). I think there is this rule of thumb also mentioned in https://doi.org/10.5194/gmd-7-1297-2014

> In order to ensure a fair evaluation,
our strategy was to always use the coarser grid as reference,
except for mean sea-level pressure 

which would mean for us:

* eobs (0.1°): regrid eobs to EURO-CORDEX grid
* ERA5: regrid EURO-CORDEX grid to ERA5

I think, the paper does not mention this, but in case of flux variables like `pr` we should use conservative remapping (i think in xesmf terms, it's "conservative_normed") and for grid point values like `tas` we can use bilinear interpolation.

and also:
> Additionally, an elevation correction was carried
out for temperature assuming a uniform temperature lapse
rate of 0.0064 K m−1

see, e.g., https://github.com/euro-cordex/evaltools/blob/b42eb5a5228872f467b6530fd70fdf8684e66a4e/evaltools/eval.py#L89-L100

**Comment by JavierDiezSierra** [comment]
@larsbuntemeyer If we decide to use bilinear interpolation, it is a one-liner with xESMF, as you mentioned. However, for the conservative remapping method, it is necessary to calculate the lon and lat boundaries of the cells. For that, we use this approximation [estimate_boundaries](https://github.com/SantanderMetGroup/python-resources/blob/main/lib/interpolater.py#L52)

**Comment by larsbuntemeyer** [comment]
Yes, we need bounds but they should be provided in the datasets as `lon_vertices` and `lat_vertices` which xESMF should understand. So we shouldnt' have to estimate them!

However, your code looks interesting, did you check [cf.add_bounds](https://cf-xarray.readthedocs.io/en/latest/bounds.html)? Should be interesting to compare!

For CORDEX rotated pole grid mappings, i have a function [transform_bounds](https://py-cordex.readthedocs.io/en/latest/generated/cordex.transform_bounds.html) that can compute bounds exactly if they are missing using pyproj. 

One problem i see is that `ALADIN` datasets don't have lon and lat bounds, but it should be possible also to compute them exactly using transform_bounds.

**Comment by larsbuntemeyer** [comment]
Here is a quick prototype how to handle bounds when regridding eobs conservatively in xesmf:

```python
import xarray as xr
import numpy as np
import cf_xarray as cfxr
from xesmf import Regridder

store = "https://ncsa.osn.xsede.org/Pangeo/pangeo-forge/pangeo-forge/EOBS-feedstock/eobs-tg-tn-tx-rr-hu-pp.zarr"
eobs = xr.open_dataset(store, engine="zarr", chunks={})
#eobs["mask"] = xr.where(~np.isnan(eobs.pp.isel(time=0)), 1, 0)

ds = xr.open_dataset("/mnt/CORDEX_CMIP6_tmp/sim_data/CORDEX/CMIP6/DD/EUR-12/GERICS/ERA5/evaluation/r1i1p1f1/REMO2020/v1-r1/mon/pr/v20241120/pr_EUR-12_ERA5_evaluation_r1i1p1f1_GERICS_REMO2020_v1-r1_mon_197901-198812.nc", decode_coords="all")

ds['lon_b'] = cfxr.bounds_to_vertices(
    ds.vertices_lon, bounds_dim="vertices", order="counterclockwise"
)
ds['lat_b'] = cfxr.bounds_to_vertices(
    ds.vertices_lat, bounds_dim="vertices", order="counterclockwise"
)

regridder = Regridder(eobs, ds, method="conservative_normed")
eobs_on_cordex_grid = regridder(eobs)
eobs_on_cordex_grid.pp.where(eobs_on_cordex_grid.pp>0).sel(time="2000-01-01T00:00:00").plot()
```
![grafik](https://github.com/user-attachments/assets/e7d31693-0c09-413f-b1dd-a7d417065c6d)

**Comment by JavierDiezSierra** [comment]
Really nice, @larsbuntemeyer!

We tried to generalize the regridding process by calculating longitude and latitude boundaries on our own through an approximation because some curvilinear simulations didn't provide this information. However, I didn't know about the [transform_bounds](https://py-cordex.readthedocs.io/en/latest/generated/cordex.transform_bounds.html) function. It is really interesting, and I will try it!

**Comment by larsbuntemeyer** [comment]
Yes, it's quite convenient! The [core code](https://github.com/euro-cordex/py-cordex/blob/c38c5f9c7a4c97c8eeb6d792a1cc3fdf39b0f483/cordex/transform.py#L166-L172) ist based on [pyproj.Transformer]. So if you have native coordinates (e.g., in a rotated crs) and the grid mapping definition, you can use that to transform coordinates in any other CRS.

**Comment by larsbuntemeyer** [comment]
> Note, that eobs has missing values over water so we should probably handle this using a [mask for regridding](https://pangeo-xesmf.readthedocs.io/en/latest/notebooks/Masking.html#Masking-and-extrapolation-in-xESMF).

Actually, we should use `unmapped_to_nan=True` to set eobs values that have no data on the cordex grid to nan. That's cleaner.

**Comment by gnikulin** [comment]
In general, just to keep in mind, remapping/regridding depends on direction, interpolation or aggregation.  If we are going from  coarser to finer grids (ERA5 -> EUR-12) or remapping between 2 grids with about the same resolution (EUR-12 and E-OBS), the bilinear interpolation should work fine. If we are going from finer to coarser grids (EUR-12 -> ERA5) we need to aggregate data. In this case, simple area-weighted remapping should work fine. The conservative remapping also works, although I'm not sure that we need to preserve fluxes for such kind of analysis.

**Comment by larsbuntemeyer** [comment]
Yes, the masking ist perfomed while opening up datasets using `sftlf`. Then [this mask is also used for regredding](https://pangeo-xesmf.readthedocs.io/en/latest/notebooks/Masking.html), however, esmf regrids nan to 0, so i masked it again after regridding. However, in my latest playing around with [eobs notebook](https://github.com/euro-cordex/joint-evaluation/blob/main/eval-book/eobs.ipynb), i realized you can also use [unmapped_to_nan = True](https://github.com/pangeo-data/xESMF/blob/30e3ecb094d39a0c6f1edb22b60d8d608a19f7d9/xesmf/frontend.py#L321-L325) when creating the regridder, it didn't make it to the papermill notebooks (i forgot as well as the heigt temperature correction!). I'll update the code in tools.py.
